### PR TITLE
chore(config): Add Pos to Config

### LINKED ISSUE #1698: Add `Pos` to `Config`
Currently `Config` type doesn't hold any position information in its fields. This makes it extremely difficult to debug and attach meaningful traces. In this issue, we want to insert tracing information in the config. That way we could see better errors:

**Current**
```bash
❯ tailcall check ./examples/jsonplaceholder.graphql
 INFO Env file: "/Users/tushar/Documents/Projects/tailcall/.env" loaded
 INFO File read: ./examples/jsonplaceholder.graphql ... ok
 WARN A new release of tailcall is available: v0.74.1 ➜ v0.74.12
 WARN To upgrade, run: brew upgrade tailcall
 WARN https://github.com/tailcallhq/tailcall/releases/tag/v0.74.12
ERROR Invalid Configuration
Caused by:
  • No base URL defined [at Post.user.@call.@http]
  • No base URL defined [at Query.posts.@http]
  • No base URL defined [at Query.user.@http]
  • No base URL defined [at Query.users.@http]
```

**Expected**
```bash
❯ tailcall check ./examples/jsonplaceholder.graphql
 INFO Env file: "/Users/tushar/Documents/Projects/tailcall/.env" loaded
 INFO File read: ./examples/jsonplaceholder.graphql ... ok
 WARN A new release of tailcall is available: v0.74.1 ➜ v0.74.12
 WARN To upgrade, run: brew upgrade tailcall
 WARN https://github.com/tailcallhq/tailcall/releases/tag/v0.74.12
ERROR Invalid Configuration
Caused by:
  • No base URL defined [at ./examples/jsonplaceholder.graphql#8:17]
```

Going forward we would be able to extract a piece of text from the source file and display it on the CLI with swiggly lines under it.

**Technical Requirements**
- Add a new type called `Pos` for each field in `Config`. This field is essentially transparent to the serializer/deserializer
  
  
  ```rust
  struct Pos<A> {
    line: u64,
    offset: u64,
    value: A
  }
  ```
- Implement `merge_right` on `Pos` in such a way that position information is correctly merged. Ensure information is stored in Pos from wherever the data is read from.

### PR DESCRIPTION
**Pull request by bnchi**
**Summary:**  
- Adds _Pos_ wrapper struct around _Config_ fields
- Introduces a new trait and macro to handle adding positional details to each field
- Minor refactors to _from_document_ and other parts of the code 

**Issue Reference(s):**  
Fixes #1698 
/claim #1698 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

**Comment by bnchi** [comment]
Waiting a review on this PR to know if it's a desired approach, once the review is done I'll update it to finalize the fix.

**Comment by tusharmath** [comment]
@bnchi I tried to use this PR but it doesn't seem to show the errors as provided in the PR.

**Comment by bnchi** [comment]
> @bnchi I tried to use this PR but it doesn't seem to show the errors as provided in the PR.

@tusharmath 

The current error handling in the codebase have a bug if a _baseURL_ is not specified in _@upstream_ all the other directives that depends on a common _baseURL_ like _@http_ for example errors out leading to this invalid configuration errors. 

```
ERROR Invalid Configuration
Caused by:
  • No base URL defined [at Post.user.@call.@http]
  • No base URL defined [at Query.posts.@http]
  • No base URL defined [at Query.user.@http]
  • No base URL defined [at Query.users.@http]
```

We discussed this on discord and if I'm not mistaken you mentioned we could clear this bug in a separate PR and add the position details to the trace. If you would like me to clear this bug also in this PR let me know, but this might make the PR harder to review.

This PR adds _Pos_ wrappers around all the Config fields. There's still some work to do here for some types but I just wanted to make sure it's a desired approach before changing all the other ones. 

For example given the original example file provided in the issue _examples/jsonplaceholder.graphql_ here's how a config look like now : 

<details>
  <summary>
    Click here to expand a config_module dbg! output
  </summary>

 ```
{
    config: Config {
        server: Pos {
            line: 2,
            column: 3,
            inner: Server {
                apollo_tracing: None,
                batch_requests: None,
                headers: Some(
                    Pos {
                        line: 2,
                        column: 23,
                        inner: Headers {
                            cache_control: None,
                            custom: [],
                            experimental: None,
                            set_cookies: None,
                            cors: Some(
                                Cors {
                                    allow_credentials: None,
                                    allow_headers: [
                                        "*",
                                    ],
                                    allow_methods: [
                                        POST,
                                        GET,
                                        OPTIONS,
                                    ],
                                    allow_origins: [
                                        "*",
                                    ],
                                    allow_private_network: None,
                                    expose_headers: [],
                                    max_age: None,
                                    vary: [
                                        "origin",
                                        "access-control-request-method",
                                        "access-control-request-headers",
                                    ],
                                },
                            ),
                        },
                    },
                ),
                global_response_timeout: None,
                hostname: None,
                introspection: None,
                pipeline_flush: None,
                port: Some(
                    Pos {
                        line: 2,
                        column: 11,
                        inner: 8000,
                    },
                ),
                query_validation: None,
                response_validation: None,
                script: None,
                showcase: None,
                vars: [],
                version: None,
                workers: None,
            },
        },
        upstream: Pos {
            line: 3,
            column: 3,
            inner: Upstream {
                allowed_headers: None,
                base_url: Some(
                    Pos {
                        line: 3,
                        column: 13,
                        inner: "http://jsonplaceholder.typicode.com",
                    },
                ),
                batch: Some(
                    Pos {
                        line: 3,
                        column: 76,
                        inner: Batch {
                            delay: 100,
                            headers: {},
                            max_size: None,
                        },
                    },
                ),
                connect_timeout: None,
                http_cache: Some(
                    Pos {
                        line: 3,
                        column: 61,
                        inner: 42,
                    },
                ),
                http_2_only: None,
                keep_alive_interval: None,
                keep_alive_timeout: None,
                keep_alive_while_idle: None,
                pool_max_idle_per_host: None,
                pool_idle_timeout: None,
                proxy: None,
                tcp_keep_alive: None,
                timeout: None,
                user_agent: None,
                dedupe: None,
            },
        },
        schema: RootSchema {
            query: Some(
                "Query",
            ),
            mutation: None,
            subscription: None,
        },
        types: {
            "Post": Pos {
                line: 22,
                column: 1,
                inner: Type {
                    fields: {
                        "body": Field {
                            type_of: "String",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "id": Field {
                            type_of: "Int",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "title": Field {
                            type_of: "String",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "user": Field {
                            type_of: "User",
                            list: false,
                            required: false,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: Some(
                                Pos {
                                    line: 27,
                                    column: 14,
                                    inner: Call {
                                        steps: Pos {
                                            line: 0,
                                            column: 0,
                                            inner: [
                                                Step {
                                                    query: Some(
                                                        "user",
                                                    ),
                                                    mutation: None,
                                                    args: {
                                                        "id": String("{{.value.userId}}"),
                                                    },
                                                },
                                            ],
                                        },
                                    },
                                },
                            ),
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "userId": Field {
                            type_of: "Int",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                    },
                    added_fields: [],
                    doc: None,
                    implements: {},
                    cache: None,
                    protected: None,
                    tag: None,
                },
            },
            "Query": Pos {
                line: 7,
                column: 1,
                inner: Type {
                    fields: {
                        "posts": Field {
                            type_of: "Post",
                            list: true,
                            required: false,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: Some(
                                Pos {
                                    line: 8,
                                    column: 17,
                                    inner: Http {
                                        base_url: None,
                                        body: None,
                                        encoding: ApplicationJson,
                                        group_by: [],
                                        headers: [],
                                        input: None,
                                        method: GET,
                                        path: "/posts",
                                        output: None,
                                        query: [],
                                    },
                                },
                            ),
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "user": Field {
                            type_of: "User",
                            list: false,
                            required: false,
                            list_type_required: false,
                            args: {
                                "id": Arg {
                                    type_of: "Int",
                                    list: false,
                                    required: true,
                                    doc: None,
                                    modify: None,
                                    default_value: None,
                                },
                            },
                            doc: None,
                            modify: None,
                            omit: None,
                            http: Some(
                                Pos {
                                    line: 10,
                                    column: 24,
                                    inner: Http {
                                        base_url: None,
                                        body: None,
                                        encoding: ApplicationJson,
                                        group_by: [],
                                        headers: [],
                                        input: None,
                                        method: GET,
                                        path: "/users/{{.args.id}}",
                                        output: None,
                                        query: [],
                                    },
                                },
                            ),
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "users": Field {
                            type_of: "User",
                            list: true,
                            required: false,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: Some(
                                Pos {
                                    line: 9,
                                    column: 17,
                                    inner: Http {
                                        base_url: None,
                                        body: None,
                                        encoding: ApplicationJson,
                                        group_by: [],
                                        headers: [],
                                        input: None,
                                        method: GET,
                                        path: "/users",
                                        output: None,
                                        query: [],
                                    },
                                },
                            ),
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                    },
                    added_fields: [],
                    doc: None,
                    implements: {},
                    cache: None,
                    protected: None,
                    tag: None,
                },
            },
            "User": Pos {
                line: 13,
                column: 1,
                inner: Type {
                    fields: {
                        "email": Field {
                            type_of: "String",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "id": Field {
                            type_of: "Int",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "name": Field {
                            type_of: "String",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "phone": Field {
                            type_of: "String",
                            list: false,
                            required: false,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "username": Field {
                            type_of: "String",
                            list: false,
                            required: true,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                        "website": Field {
                            type_of: "String",
                            list: false,
                            required: false,
                            list_type_required: false,
                            args: {},
                            doc: None,
                            modify: None,
                            omit: None,
                            http: None,
                            call: None,
                            grpc: None,
                            script: None,
                            const_field: None,
                            graphql: None,
                            cache: None,
                            protected: None,
                        },
                    },
                    added_fields: [],
                    doc: None,
                    implements: {},
                    cache: None,
                    protected: None,
                    tag: None,
                },
            },
        },
        unions: {},
        enums: {},
        links: [],
        telemetry: Pos {
            line: 0,
            column: 0,
            inner: Telemetry {
                export: None,
                request_headers: [],
            },
        },
    },
    extensions: Extensions {
        grpc_file_descriptors: {},
        script: None,
        cert: [],
        keys: [],
        endpoint_set: EndpointSet {
            endpoints: [],
            marker: PhantomData<tailcall::core::rest::endpoint_set::Unchecked>,
        },
        htpasswd: [],
        jwks: [],
    },
    input_types: {},
    output_types: {
        "Query",
        "Post",
        "User",
    },
    interface_types: {},
}
```
</details>

**Review by tusharmath — COMMENTED** [review]
I think the trace still needs fixing. The error is repeated, which is fine and can be fixed separately, but the trace produced currently is quite messy and confusing.

**Comment by tusharmath** [comment]
@bnchi Moving it to draft. Whenever you are ready and the test case mentioned in the issue passes, please mark it as ready to review.

**Comment by bnchi** [comment]
The PR is ready for a review, the lint and darwin-arm64 are failing in the CI for some reason there's no issues or errors but the build just can't seem to continue on those two checks.

@tusharmath

**Comment by bnchi** [comment]
I noticed a bug the new _Pos_ wrapper type breaks gen_gql_schema.rs it seems like there are still work that needs to be done there in order for these changes to not break the generation of the types in that file, I'm going to bring this back to a draft.

File: src/core/config/config.rs [hunk 1]
```diff
@@ -694,19 +828,19 @@ impl Config {
...
 
     pub fn from_source(source: Source, schema: &str) -> Result<Self> {
-        match source {
-            Source::GraphQL => Ok(Config::from_sdl(schema).to_result()?),
-            Source::Json => Ok(Config::from_json(schema)?),
-            Source::Yml => Ok(Config::from_yaml(schema)?),
+        match source.input_type {
+            SourceType::GraphQL => Ok(Config::from_sdl(&source.input_path, schema).to_result()?),
+            SourceType::Json => Ok(Config::from_json(schema)?),
+            SourceType::Yml => Ok(Config::from_yaml(schema)?),
```
Comment (tusharmath) [inline src/core/config/config.rs:843]: The position information should be available in `Json` and `YML` format also. This is what is happenening right now

```bash
cargo run -- start ./examples/jsonplaceholder.json
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.73s
     Running `target/debug/tailcall start ./examples/jsonplaceholder.json`
 INFO File read: ./examples/jsonplaceholder.json ... ok
 INFO N + 1 detected: 1
ERROR Invalid Configuration
Caused by:
  • No base URL defined [at  0#0]
  • No base URL defined [at  0#0]
  • No base URL defined [at  0#0]
  ```

File: tailcall-macros/src/positioned_config.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+                Ok(())
+            })?;
+        }
+    }
+
+    Ok(parsed_attr)
+}
+
+pub fn expand_positoned_config(item: TokenStream) -> TokenStream {
```
Comment (tusharmath) [inline tailcall-macros/src/positioned_config.rs:62]:
```suggestion
pub fn expand_positioned_config(item: TokenStream) -> TokenStream {
```

File: src/core/config/config.rs (hunk 1 above)
Comment (bnchi) [reply src/core/config/config.rs:843]: Yeah that's bad, I guess I have to find a way to get the positional details for when serde deserialize the data from json and yaml

**Comment by bnchi** [comment]
Getting the line and column from JSON and YML isn't an easy task a lot of the serde implementations doesn't seem to provide a way to get those from source, even if it was possible this will cause a complete shift on how to approach this task. For those reasons I'm going to close the PR

**Comment by bnchi** [comment]
The PR is ready for review here are the latest changes summary : 

- Introduced a new field to _Cause_ to store something of type _SourcePos_ that allow us to identify the position at which an error occurred from a file 
- Added a new method to Validator trait that can be called on _Valid_ which helps decouple the _trace_ from positioned errors so now we can just call this method for places that support these types of errors
- _CLIError_ struct now as well got a new field to show positioned errors in the console in case it was supported. When printing we favor positioned error messages over normal traces
- Refactored _json_schema_  and added a new type that allow us to point exactly where an error occurred from source when comparing two schemas together that are read from source  

I would appreciate if you can closely review the changes made to the new _Transformer_ logic and test that because while developing this PR I had to deal with a lot of new changes there while developing this feature, so smoke/integration testing to make sure everything works.

**Pull request closed without merging**